coderd/subdomain.go

```diff
@@ -0,0 +1,155 @@
+package coderd
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/coder/coder/coderd/httpapi"
+
+	"github.com/coder/coder/coderd/httpmw"
+
+	"github.com/go-chi/chi/v5"
+
+	"golang.org/x/xerrors"
+)
+
+const (
+	// XForwardedHostHeader is a header used by proxies to indicate the
+	// original host of the request.
+	XForwardedHostHeader = "X-Forwarded-Host"
+)
+
+// ApplicationURL is a parsed application url into it's components
+type ApplicationURL struct {
+	AppName       string
+	WorkspaceName string
+	Agent         string
+	Username      string
+	Path          string
+	Domain        string
+}
+
+func (api *API) handleSubdomain(middlewares ...func(http.Handler) http.Handler) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+			app, err := ParseSubdomainAppURL(r)
+
+			if err != nil {
+				// Subdomain is not a valid application url. Pass through.
+				// TODO: @emyrk we should probably catch invalid subdomains. Meaning
+				// 	an invalid application should not route to the coderd.
+				//	To do this we would need to know the list of valid access urls
+				//	though?
+				next.ServeHTTP(rw, r)
+				return
+			}
+
+			workspaceAgentKey := app.WorkspaceName
+			if app.Agent != "" {
+				workspaceAgentKey += "." + app.Agent
+			}
+			chiCtx := chi.RouteContext(ctx)
+			chiCtx.URLParams.Add("workspace_and_agent", workspaceAgentKey)
+			chiCtx.URLParams.Add("user", app.Username)
+
+			// Use the passed in app middlewares before passing to the proxy app.
+			mws := chi.Middlewares(middlewares)
+			mws.Handler(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+				workspace := httpmw.WorkspaceParam(r)
+				agent := httpmw.WorkspaceAgentParam(r)
+
+				api.proxyWorkspaceApplication(proxyApplication{
+					Workspace: workspace,
+					Agent:     agent,
+					AppName:   app.AppName,
+				}, rw, r)
+			})).ServeHTTP(rw, r.WithContext(ctx))
+		})
+	}
+}
+
+var (
+	// Remove the "starts with" and "ends with" regex components.
+	nameRegex = strings.Trim(httpapi.UsernameValidRegexString(), "^$")
+	appURL    = regexp.MustCompile(fmt.Sprintf(
+		// {USERNAME}--{WORKSPACE_NAME}}--{{AGENT_NAME}}--{{PORT}}
+		`^(?P<UserName>%[1]s)--(?P<WorkspaceName>%[1]s)(--(?P<AgentName>%[1]s))?--(?P<AppName>%[1]s)$`,
+		nameRegex))
+)
+
+// ParseSubdomainAppURL parses an application from the subdomain of r's Host header.
+// If the application string is not valid, returns a non-nil error.
+//   1) {USERNAME}--{WORKSPACE_NAME}}--{{AGENT_NAME}}--{{PORT/AppName}}
+//  	(eg. http://admin--myenv--main--8080.cdrdeploy.c8s.io)
+func ParseSubdomainAppURL(r *http.Request) (ApplicationURL, error) {
```

Comment:
From my perspective, this parsing logic should entirely be placed in the `httpmw` package instead. It'll be a bit simpler to test, won't need to be exported since it'll be in the handler, and probably be small. Thoughts?

Reply:
I don't think I follow. Do you mean like an `ExtrackSubdomainAppURL` method to pull it out in the handler?